### ISSUE TITLE
Improvement - WC_Admin_Report::get_order_report_data() JOIN clauses

### ISSUE THREAD

**Issue by daigo75** [woothemes/woocommerce#7818]
Method `WC_Admin_Report::get_order_report_data()`, at the moment, only supports `INNER JOIN` clauses, even though it uses `LEFT JOIN` in the SQL.

Example
--
Consider the following call:
~~~
$data_query = array(
	'data' => array(
		'ID' => array(
			'type'     => 'post_data',
			'function' => '',
			'name'     => 'post_id'
		),
		// Suppose that "_some_meta" may or may not exist 
		'_some_meta' => array(
			'type' => 'meta',
			'order_item_type' => '',
			'function' => '',
			'name' => '_some_meta'
		)
	)
);
~~~

It produces the following SQL (reformatted for clarity):

~~~
SELECT
  posts.ID as post_id
  ,meta__some_meta.meta_value as _some_meta
FROM
  wp_posts AS posts 
  LEFT JOIN 
  wp_postmeta AS meta__some_meta ON 
    (posts.ID = meta__some_meta.post_id)
WHERE
  posts.post_type IN ( 'shop_order' )		
  AND posts.post_status 	IN ( 'wc-processing','wc-completed')
  AND posts.post_date >= '2014-10-01'
  AND posts.post_date < '2015-01-01'
  -- This clause belongs in the LEFT JOIN clause
  AND meta__some_meta.meta_key = '_some_meta'
~~~

The SQL won't return any order if it doesn't have a `_some_meta` meta attribute, due to the `WHERE` clause at the bottom. However, the `LEFT JOIN` clause should be used when the order data **should** be returned whether the meta key exists or not. That is, the above query is a more confusing alternative of the following:

~~~
SELECT
  posts.ID as post_id
  ,meta__some_meta.meta_value as _some_meta
FROM
  wp_posts AS posts 
  INNER JOIN 
  -- Both clauses in the INNER JOIN. No order will be returned if the meta is not present
  wp_postmeta AS meta__some_meta ON 
    (posts.ID = meta__some_meta.post_id) AND
    (meta__some_meta.meta_key = '_some_meta')
WHERE
  posts.post_type IN ( 'shop_order' )		
  AND posts.post_status IN ( 'wc-processing','wc-completed')
  AND posts.post_date >= '2014-10-01'
  AND posts.post_date < '2015-01-01'
~~~

However, this is not always the desired result. One would probably want to return the order whether is has the meta attribute or not. In such case, the query should be the following:

~~~
SELECT
  posts.ID as post_id
  ,meta__some_meta.meta_value as _some_meta
FROM
  wp_posts AS posts 
  LEFT JOIN 
  -- Both clauses in the LEFT JOIN. Orders without the meta key will
  -- be returned, and the meta_value for _some_meta will simply be NULL
  wp_postmeta AS meta__some_meta ON 
    (posts.ID = meta__some_meta.post_id) AND
    (meta__some_meta.meta_key = '_some_meta')
WHERE
  posts.post_type IN ( 'shop_order' )		
  AND posts.post_status IN ( 'wc-processing','wc-completed')
  AND posts.post_date >= '2014-10-01'
  AND posts.post_date < '2015-01-01'
~~~

Suggestion
--
1. Modify `WC_Admin_Report::get_order_report_data()` so that it puts all the meta clauses in the `JOIN`, and replace `LEFT JOIN` with `INNER JOIN`
2. Add support for a `join_type` argument, so that the `INNER JOIN` can be modified to a `LEFT` or `RIGHT JOIN` if required.

Proof of concept
--
See gist: https://gist.github.com/daigo75/2f88fa4d1b0907e080cd

**Comment by mikejolley** [comment]
@daigo75 This makes sense, and the gist looks good. If you do a pull request on master I'll merge it in and you'll get props. Thanks. :+1:

**Comment by daigo75** [comment]
Done. As always, the changes are fully backward compatible, and anyone who doesn't use the new `join_type` parameter should not see any difference (the queries are just slightly different, as they use the `INNER JOIN`).

Of course, I would recommend testing the changes in your environments, as the original logic used to populate the `WHERE` was a bit confusing and it wasn't easy to refactor it. :sweat_drops:

**Comment by mikejolley** [comment]
Closed in favour of #7836
